RecoMET/METPUSubtraction/plugins/PFMETProducerMVA.cc

```diff
@@ -241,58 +129,109 @@ void PFMETProducerMVA::produce(edm::Event& evt, const edm::EventSetup& es)
+	    }
+	  }
+	  if(pMatch) break;
+	}
+	if(pMatch) break;
+      }
+      if(pMatch) continue;
+      reco::PUSubMETCandInfo pLeptonInfo;
+      pLeptonInfo.p4_          = lepton1->p4();
+      pLeptonInfo.chargedEnFrac_ = chargedEnFrac(&(*lepton1),pfCandidates_view,hardScatterVertex);
+      leptonInfo.push_back(pLeptonInfo); 
+      if(lepton1->isPhoton()) { lHasPhotons = true; }
+    }
+    lId++;
+  }
+ 
+  return leptonInfo;
+}
 
+
+std::vector<reco::PUSubMETCandInfo> 
+PFMETProducerMVA::computeJetInfo(const reco::PFJetCollection& uncorrJets, 
+				 const edm::Handle<reco::PFJetCollection>& corrJets, 
+				 const edm::ValueMap<float>& jetIds,
+				 const reco::VertexCollection& vertices,
+				 const reco::Vertex* hardScatterVertex,
+				 const JetCorrector &iCorrector,edm::Event &iEvent,const edm::EventSetup &iSetup,
+				 std::vector<reco::PUSubMETCandInfo> &iLeptons,std::vector<reco::PUSubMETCandInfo> &iCands)
+{
   const L1FastjetCorrector* lCorrector = dynamic_cast<const L1FastjetCorrector*>(&iCorrector);
-  std::vector<mvaMEtUtilities::JetInfo> retVal;
+  std::vector<reco::PUSubMETCandInfo> retVal;
   for ( reco::PFJetCollection::const_iterator uncorrJet = uncorrJets.begin();
 	uncorrJet != uncorrJets.end(); ++uncorrJet ) {
-    //int pIndex = uncorrJet-uncorrJets.begin();
-    //edm::RefToBase<reco::Jet> jetRef(edm::Ref<PFJetCollection>(&uncorrJets,pIndex));
-    for ( reco::PFJetCollection::const_iterator corrJet = corrJets.begin();
-	  corrJet != corrJets.end(); ++corrJet ) {
+    // for ( reco::PFJetCollection::const_iterator corrJet = corrJets.begin();
+    // 	  corrJet != corrJets.end(); ++corrJet ) {
+    for( size_t cjIdx=0;cjIdx<corrJets->size();cjIdx++) {
+      reco::PFJetRef corrJet( corrJets, cjIdx );
+
       // match corrected and uncorrected jets
       if ( uncorrJet->jetArea() != corrJet->jetArea() ) continue;
       if ( deltaR2(corrJet->p4(),uncorrJet->p4()) > dR2Min ) continue;
 
       // check that jet passes loose PFJet id.
-      //bool passesLooseJetId = (*looseJetIdAlgo_)(*corrJet);
-      //if ( !passesLooseJetId ) continue; 
       if(!passPFLooseId(&(*uncorrJet))) continue;
 
       // compute jet energy correction factor
       // (= ratio of corrected/uncorrected jet Pt)
-      double jetEnCorrFactor = corrJet->pt()/uncorrJet->pt();
-      mvaMEtUtilities::JetInfo jetInfo;
+      //double jetEnCorrFactor = corrJet->pt()/uncorrJet->pt();
+      reco::PUSubMETCandInfo jetInfo;
       
       // PH: apply jet energy corrections for all Jets ignoring recommendations
       jetInfo.p4_ = corrJet->p4();
       double lType1Corr = 0;
       if(useType1_) { //Compute the type 1 correction ===> This code is crap 
-	double pCorr = lCorrector->correction(*uncorrJet,iEvent,iSetup); //Does not work in 42X 
+	double pCorr = lCorrector->correction(*uncorrJet,iEvent,iSetup);
 	lType1Corr = (corrJet->pt()-pCorr*uncorrJet->pt());
```

Comment:
does it mean that all corrections increase jet momentum?
I'd still suggest to take an abs or put an assert here if it really shouldn't happen.